v1/common/broker.go

```diff
@@ -137,3 +137,101 @@ func (b *Broker) AdjustRoutingKey(s *tasks.Signature) {
 
 	s.RoutingKey = b.GetConfig().DefaultQueue
 }
+
+type resizableCapacity struct {
+	changes chan change
+
+	pool chan struct{}
+}
+
+type change struct {
+	isReturned bool
+	updatedCap *int
+}
+
+func NewResizablePool(startingCapacity int) (iface.ResizeablePool, func()) {
+	rc := &resizableCapacity{
+		changes: make(chan change),
+		pool:    make(chan struct{}),
+	}
+
+	var (
+		// A separate routine is used to push any available capacity to the pool, when available. This channel allows
+		// cancelling the last function when available capacity changes (both to avoid leaking resources and so that
+		// capacity available at one point that wasn't used isn't given when capacity is lowered.
+		lastGiverCancel chan struct{}
+
+		capacity int
+		taken    int
+
+		// Used to cancel the background routine/release resources
+		cancelChan = make(chan struct{}, 1)
+	)
+
+	// Use a routine to coordinate changes (either returned capacity or changed capacity)
+	// This allows the different paths to both ultimately control a single channel that can be used as a normal
+	// capacity pool channel.
+	go func() {
+		for {
+			select {
+			case <-cancelChan:
+				if lastGiverCancel != nil {
+					lastGiverCancel <- struct{}{}
+				}
+
+				return
+			case c := <-rc.changes:
+				// Kill the function that might be giving capacity -- a new function will be started if any is available
+				if lastGiverCancel != nil {
+					lastGiverCancel <- struct{}{}
+					lastGiverCancel = nil
+				}
+
+				// Calculate what's available based on the change
+				if c.isReturned {
+					taken--
```

Comment:
This isn't used by multiple threads. `rc.changes` ensures that capacity/counts changes can come from any threads but only be processed on the single thread/go routine.

It's generally encouraged in go to use channels to pass messages rather than lock resources.